src/linear_algebra/quadratic_form.lean

```diff
@@ -0,0 +1,290 @@
+⟨ λ a Q,
+  { map := λ x, Q (a • x),
+    map_smul := λ b x, by rw [smul_comm, apply_smul],
+    polar_add_left := λ x x' y,
+      by convert Q.polar_add_left (a • x) (a • x') (a • y) using 1; simp [polar, smul_add],
+    polar_smul_left := λ b x y,
+      by convert Q.polar_smul_left b (a • x) (a • y) using 1; simp [polar, smul_add, smul_comm],
+    polar_add_right := λ x y y',
+      by convert Q.polar_add_right (a • x) (a • y) (a • y') using 1; simp [polar, smul_add],
+    polar_smul_right := λ b x y,
+      by convert Q.polar_smul_right b (a • x) (a • y) using 1; simp [polar, smul_add, smul_comm] } ⟩
+
+@[simp] lemma smul_apply (a : R₁) (Q : quadratic_form R₁ M) (x : M) : (a • Q) x = Q (a • x) := rfl
+
+instance : mul_action R₁ (quadratic_form R₁ M) :=
+{ mul_smul := λ a b Q, ext (λ x, by simp [mul_smul, smul_comm]),
+  one_smul := λ Q, ext (λ x, by simp),
+  ..quadratic_form.has_scalar }
+
+section comp
+
+variables {N : Type v} [add_comm_group N] [module R N]
+
+/-- Compose the quadratic form with a linear function. -/
+def comp (Q : quadratic_form R N) (f : M →ₗ[R] N) :
+  quadratic_form R M :=
+{ map := λ x, Q (f x),
+  map_smul := λ a x, by simp [apply_smul],
+  polar_add_left := λ x x' y,
+    by convert Q.polar_add_left (f x) (f x') (f y) using 1; simp [polar],
+  polar_smul_left := λ a x y,
+    by convert Q.polar_smul_left a (f x) (f y) using 1; simp [polar],
+  polar_add_right := λ x y y',
+    by convert Q.polar_add_right (f x) (f y) (f y') using 1; simp [polar],
+  polar_smul_right := λ a x y,
+    by convert Q.polar_smul_right a (f x) (f y) using 1; simp [polar] }
+
+@[simp] lemma comp_apply (Q : quadratic_form R N) (f : M →ₗ[R] N) (x : M) :
+  (Q.comp f) x = Q (f x) := rfl
+
+end comp
+
+end quadratic_form
+
+/-! ### Associated bilinear forms
+
+Over a commutative ring where 2⁻¹ exists, the theory of quadratic forms is
+basically identical to that of symmetric bilinear forms. The map from quadratic
+forms to bilinear forms giving this identification is called the `associated`
+quadratic form.
+-/
+
+variables {B : bilin_form R M}
+
+namespace bilin_form
+open quadratic_form
+
+lemma polar_to_quadratic_form (x y : M) : polar (λ x, B x x) x y = B x y + B y x :=
+by simp [polar, add_left, add_right, sub_eq_add_neg _ (B y y), add_comm (B y x) _]
+
+/-- A bilinear form gives a quadratic form by applying the argument twice. -/
+def to_quadratic_form (B : bilin_form R M) : quadratic_form R M :=
+⟨ λ x, B x x,
+  λ a x, by simp [smul_left, smul_right, mul_assoc],
+  λ x x' y, by simp [polar_to_quadratic_form, add_left, add_right, add_left_comm],
+  λ a x y, by simp [polar_to_quadratic_form, smul_left, smul_right, mul_add],
+  λ x y y', by simp [polar_to_quadratic_form, add_left, add_right, add_left_comm],
+  λ a x y, by simp [polar_to_quadratic_form, smul_left, smul_right, mul_add] ⟩
+
+@[simp] lemma to_quadratic_form_apply (B : bilin_form R M) :
+(B.to_quadratic_form : M → R) = λ x, B x x :=
```

Comment:
Is there a reason why you write this as a lambda, instead of putting `x` as a hypothesis before the colon?

Reply:
Not really. Fixed!